bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/Future.java

```diff
@@ -26,13 +31,19 @@
  */
 public class Future {
     private final Strand strand;
+    private final AtomicBoolean visited = new AtomicBoolean(false);
 
      Future(Strand strand) {
         this.strand = strand;
     }
 
     public void complete(Object returnValue) {
+        if (visited.get()) {
```

Comment:
let's use `compareAndSet` or `getAndSet`

Reply:
What happens if I use markAsync twice which returns two futures and try to complete two futures. Still, strand will get unblocked twice.

Reply:
strand.lock();
        if (strand.getState() != State.BLOCK_AND_YIELD) {
             throw ErrorCreator.createError(StringUtils.fromString("cannot complete the same future twice."),
                                           new MapValueImpl<>(PredefinedTypes.TYPE_ERROR_DETAIL));
         }
         strand.unlock();

Reply:
Good point. Maybe we can do the same fix for `markAsync` as well.

Reply:
I made the change suggested by @warunalakshitha . @manuranga I don't think we need to change `markAsync` WDYT?

Reply:
I see you have removed the atomic boolean and used a lock, isn't this slower?

Reply:
This is to prevent the strand state changing before the  if check completes.

Reply:
this is an edge case, should we slow down the whole thing to catch this error?

Reply:
If we depend on atomic boolean inside the future object, if we have multiple futures still strand can be unblocked multiple times. We already have a lock inside unblock strand method. If this is slow, we can have another method called checkAndUnblockStarnd inside the scheduler and do the check there.

https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Scheduler.java#L422

Reply:
IMO its good to have another check-in markAsync as manu said. Since these are user apis, its good to have validations which prevent unnecessary states in the scheduler.

Reply:
Made the future in `markAsync` as an instance variable and used atomic boolean in the future